### DISCUSSION TITLE
Could we have access to BPFMap private fields?

### DISCUSSION THREAD

**Discussion by geyslan** [aquasecurity/libbpfgo#40] [Ideas]
Once BPFMap already maintains private fields which could be needed further in the code, could the consumer have access to them? E.g.:

```go
func (b *BPFMap) CloseObject() {
  b.module.Close()
}

func (b *BPFMap) GetName() (string) {
  return b.name
}

func (b *BPFMap) GetFD() (int) {
  return int(b.fd)
}
```

**Comment by yanivagman** [comment]
I don't see a reason why not to have GetName() and GetFd(), but I don't think that a map should be able to close the module itself as this should be done with the module object.
You are welcome to open a PR for this.

**Reply by geyslan** [reply]
> I don't think that a map should be able to close the module itself as this should be done with the module object.

I suppose that closing itself could provide de-duplication of the Module object for the consumer. The consumer would store only the `*BPFMap` after the `Module.BPFLoadObject()` call.

> You are welcome to open a PR for this.

Sure thing, thanks. Please, let me know about the `CloseObject()` (or `CloseModule()`) suggestion too.

P.S.:
The suggestions could scale to `BPFProg`.

**Reply by rafaeldtinoco** [reply]
> > I don't think that a map should be able to close the module itself as this should be done with the module object.
> 
> I suppose that closing itself could provide de-duplication of the Module object for the consumer. The consumer would store only the `*BPFMap` after the `Module.BPFLoadObject()` call.

You mean for BPFModule that has single maps ? That won't always be the case AND might cause problems if there are multiple maps. For example: closure would have to check if other maps exist and if they do, if they are attached.

That is not what libbpf does to bpf_object (which is closest thing we have to BPFModule):

```
	struct bpf_program *programs;
	size_t nr_programs;
	struct bpf_map *maps;
	size_t nr_maps; <- would need to be checked, and if the maps are mmaped, etc...
	size_t maps_cap;
```

**Reply by geyslan** [reply]
> You mean for BPFModule that has single maps ?

Yes, exactly that. A single map inside a single ebpf object.

> For example: closure would have to check if other maps exist and if they do, if they are attached.

Current implementation of the `Module.Close()` is just calling `bpf_object__close()` without taking care of maps. So I got your concerns. Checking the other maps existence would be necessary for that.

**Reply by rafaeldtinoco** [reply]
> Current implementation of the `Module.Close()` is just calling `bpf_object__close()` without taking care of maps. So I got your concerns. Checking the other maps existence would be necessary for that.

Yes, you are right. libbpfgo is basically a wrapper around libbpf. By calling Module.Close() we are calling bpf_object_close() and then bpf_link__destroy(). That will call call bpf_map__destroy() for all existing maps for that bpf_object. 

It is weird for me that a map is capable of calling the object to cleanup itself and all other existing maps (thus my concern, which is also not addressed by libbpf, because calling link destroy should be made separately).

I'm in favor of explicitly making consumer to call Close from the link only by the Module (and object closure).

**Reply by geyslan** [reply]
> It is weird for me that a map is capable of calling the object to cleanup itself and all other existing maps (thus my concern, which is also not addressed by libbpf, because calling link destroy should be made separately).
> I'm in favor of explicitly making consumer to call Close from the link only by the Module (and object closure).

On second thought, I agree with you.

**Comment by rafaeldtinoco** [comment]
And could you also give an example on why getting BPF name and FD could be good ? That could help others.

**Reply by itaysk** [reply]
+1 could you describe the use case? maybe instead or in addition to exporting this information we could provide a function that fulfilled the use case you had

**Reply by yanivagman** [reply]
I can see why one want to know about the name of the map later in the code.
For exame, say that you save an array of BPFMaps and you close/update/do something with it. Then it might be usefull to print/log the name of the map that you did something with.
Not sure about the fd usecase though.

**Reply by geyslan** [reply]
@rafaeldtinoco @itaysk @yanivagman 

Sure. One might want to store FD to feed other maps entries, depending on the software whole logic. Or one might just want to recover this FD and feed other software that doesn't make use of `libbpfgo`.

> For exame, say that you save an array of BPFMaps and you close/update/do something with it. Then it might be usefull to print/log the name of the map that you did something with.

Exactly, @yanivagman. This is also related to the `CloseXXX()` suggestion.

**Reply by rafaeldtinoco** [reply]
> @rafaeldtinoco @itaysk @yanivagman
> 
> Sure. One might want to store FD to feed other maps entries, depending on the software whole logic. Or one might just want to recover this FD and feed other software that doesn't make use of `libbpfgo`.

To get the MAP file-descriptor from other software you would either have to pin it (check bpftool map pin) or message it through [something like this](https://github.com/rafaeldtinoco/howtos/blob/main/codes/passfd.c). No ?

**Reply by geyslan** [reply]
> To get the MAP file-descriptor from other software you would either have to pin it (check bpftool map pin) or message it through something like this. No ?

It's a way. The maps in question are being pinned, but I think it would be nice to get the FD programmatically from the lib used to pin it. What about to provide the pinpath instead?

**Reply by rafaeldtinoco** [reply]
Yes, to implement BPFMap->GetPinPath() wouldn't be bad.
